src/lib/Microsoft.Health.Fhir.Ingest.Template/FhirLookupTemplate.cs

```diff
@@ -16,12 +16,22 @@ public class FhirLookupTemplate : ILookupTemplate<IFhirTemplate>
         public FhirLookupTemplate RegisterTemplate(IFhirTemplate fhirTemplate)
         {
             EnsureArg.IsNotNull(fhirTemplate, nameof(fhirTemplate));
-            if (_templates.ContainsKey(fhirTemplate.TypeName))
+
+            if (!string.IsNullOrWhiteSpace(fhirTemplate.TypeName))
             {
-                throw new InvalidOperationException($"Duplicate template defined for {fhirTemplate.TypeName}");
+                if (!_templates.ContainsKey(fhirTemplate.TypeName))
+                {
+                    _templates.Add(fhirTemplate.TypeName, fhirTemplate);
+                }
+                else
+                {
+                    throw new InvalidTemplateException($"Duplicate template defined for type name: '{fhirTemplate.TypeName}'");
+                }
+            }
+            else if (fhirTemplate.TypeName?.Trim().Length == 0)

```

Comment:
The first "if" covered the case when the typeName is not null and not empty. We check the map of templates under this case. For the "else" case, I wanted to alert on the "empty" case but slip on the null case, since the error message of the null case already be covered presumably by the serializer. 

This actually could be improved by a more sophisticated schema validation framework.  The newtonsoft uses Json schema to define more validations. However I don't want to create two copies of schema boilerplates. So one way I think we can do in the future is to utilize the attributes with a validation rules' engine for easily execute these logics.

Reply:
I believe Nate is referring to the string.IsNullOrWhiteSpace(fhirTemplate.TypeName) and else if (fhirTemplate.TypeName?.Trim().Length == 0).  IsNullOrWhiteSpace also checks if the string is empty.  The else if can be just an else.  *If* the else if is need then we would other logical conditions that aren't accounted for in the current code which also isn't correct.  Reiterating Nate's suggestion of replacing the else if with an else.

Reply:
Updated the code to remove that unnecessary check.